ci/clang-and-tidy.sh

```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+
+###
+# This script is a wrapper around clang and clang-tidy. It behaves mostly like
+# clang (or clang++, depending on the program name), but also executes
+# clang-tidy on the given source file.
+#
+# The purpose of this is to reduce CI check times by tricking ccache into
+# serving as a cache for clang-tidy as well. The base assumption for this is
+# inputs for clang-tidy match those of the compiler, so a ccache hit implies
+# that this file has passed the clang-tidy checks before.
+#
+# For simplicity, this script makes a couple of assumptions, that hold within
+# the our controlled CI environment:
+#   * there is only ever one source file per compiler invocation
+#   * the source file is always the last argument
+#   * the source file name matches *.cpp
+#   * the compile_commands.json in the build dir corresponds the current build
+#   * clang-tidy gets its config string passed via -config=
+#   * if a different clang or clang-tidy is used, this script is also modified
+#     (because ccache wouldn't detect any change of mtimes)
+###
+
+set -o errexit
+# set -o nounset
+
+# Perform the regular compiler invocation:
+progname=$(basename "$0")
+if [[ $progname == clang-* ]]; then
+    clang-9 "$@"
+elif [[ $progname == clang++-* ]]; then
+    clang++-9 "$@"
```

Comment:
Seeing that we will maybe switch to clang-10 soon, maybe extract CLANG_VERSION variable?

Reply:
Well, this is tricky. See the comment at the top: In a normal setup, ccache knows to separate cache entries made with different compilers based on the mtime of the compiler binary. With clang-and-tidy.sh, we're using a content hash instead of the mtime. That still means that we need the content to change when the compiler changes. Putting the program name with its version in the file is an approximation of that. Ideally, we'd put the compiler binary's hash in the script and verify it every time.

Btw, this reminds me that I'll still need to handle the .clang-tidy file. Right now, changing the config doesn't make it check everything again (which it shoud).

Reply:
Alright, this is fixed now, so the PR should be ready I think.